metaflow/cli.py

```diff
@@ -118,6 +120,26 @@ def logger(body="", system_msg=False, head="", bad=False, timestamp=True, nl=Tru
     click.secho(body, bold=system_msg, fg=LOGGER_BAD_COLOR if bad else None, nl=nl)
 
 
+def use_conf_and_merge_cb(ctx, param, value):
+    # Callback to:
+    #  - read  the Click auto_envvar variable from both the
+    #    environment AND the configuration
+    #  - merge that value with the value passed in the command line (value)
+    #  - return the value as a tuple
+    # Note that this function gets called even if there is no option passed on the
+    # command line.
+    # NOTE: Assumes that ctx.auto_envvar_prefix is set to METAFLOW (same as in
+    # from_conf)
+
+    # Special case where DECOSPECS and value are the same. This happens
+    # when there is no --with option at the TL and DECOSPECS is read from
+    # the env var. In this case, click also passes it as value
+    splits = DECOSPECS.split()
+    if len(splits) == len(value) and all([a == b for (a, b) in zip(splits, value)]):
+        return value
+    return tuple(DECOSPECS.split() + list(value))
```

Comment:
Maybe this was decided already before, but noting some oddities regarding the hierarchy in the config.

test setup:
```
# metaflow config.json
"METAFLOW_DECOSPECS": "kubernetes:memory=3072"

# cmd
python HelloFlow.py run --with "kubernetes:memory=1024,cpu=2"
```
results in the higher memory value being used, and cpu omitted.

--- 
same setup with additional env var set
```
METAFLOW_DECOSPECS="kubernetes:memory=2048"
```
results in env var overriding value from config.json (as expected?)

---
directly supplying values in a deco seems to override everything else as expected
```python
@kubernetes(memory=512, cpu=3)
```

---

so the succession right now seems to be 
```
explicit decorator in flow file
> METAFLOW_DECOSPEC env
> METAFLOW_DECOSPEC in config.json
> CLI --with
```

Reply:
the only thing I found odd with the succession is treating an explicit CLI `--with` secondary when the config/envvar values are advertised as setting defaults only. My expectation would be that an option passed through CLI would perhaps override defaults

Disregard if this is intended behaviour :)

Reply:
This is indeed the behavior that is coded. There is no notion of "decorator merge" although there should be. For now though, I think this is good enough but going forward we should see how we can merge decorators. This is the same problem as the @environment decorator where one overrides all others (and they don't merge). There is also no warning of duplicate decorators because they are all non static. In short, we have not changed the behavior at all but it can be a bit weird. You can try for example `--with kubernetes:memory=2048 --with kubernetes:memory=1024,cpu=2`, the second one will be ignored just like it is right now. You can view the default decospecs as stuff that is prepended basically. Am happy to make it post pended though if ppl prefer, as in:
explicit deco in flow file
CLI --with
METAFLOW_DECOSPEC env
METAFLOW_DECOSPEC in config.json.

Given we have no promises on any ordering right now, I don't view that as a huge issue and I think we need to more cleanly address it but that would be for another set of changes.